### DISCUSSION TITLE
How is feature matrix mapped to nodes?

### DISCUSSION THREAD

**Discussion by SaffronWolf** [pyg-team/pytorch_geometric#4304] [Q&A] [answered]
Sorry for the silly question, but when initializing `Data` object, how PyG knows what row in feature matrix belongs to which node?

```
import torch
from torch_geometric.data import Data

edge_index = torch.tensor([[0, 1],
                           [1, 0],
                           [1, 2],
                           [2, 1]], dtype=torch.long)
x = torch.tensor([[-1], [0], [1]], dtype=torch.float)

data = Data(x=x, edge_index=edge_index.t().contiguous())
```

In the above example, how do I ensure that.

[-1] is mapped to node 0
[0] is mapped to node 1
[1] is mapped to node 2

**Comment by rusty1s** [comment, marked as answer]
The node indices correspond to the row index of the feature matrix. That is, `x[0]` denotes the features of node `0`, `x[1]` denotes the features of node `1`, and so on. As such, node indices in PyG need to be mapped to the `0` to `num_nodes - 1` range. Let me know if this resolves your issue :)

**Comment by SaffronWolf** [comment]
Thanks a lot. Just curious, what happens if node indices are _not_ mapped to `0` to `num_nodes -1`. How would feature matrix be mapped to nodes then?

**Reply by rusty1s** [reply]
Interesting question, although not sure I fully understand it. In case you allow arbitrary node indices, you would need to store feature vectors in a hash map rather than in a single feature matrix, which complicates memory access patterns and slows down neural network execution.
